mixnet/client.py

```diff
@@ -1,76 +1,119 @@
+        and schedules redundant real packets to be emitted in the next turns.
+
+        If no real packet is not scheduled, this thread emits a cover packet according to the emission_rate_per_min.
+        """
+
+        redundant_real_packet_queue: PacketQueue = asyncio.Queue()
+
+        emission_notifier_queue = asyncio.Queue()
+        _ = asyncio.create_task(
+            self.__emission_notifier(
+                self.__config.emission_rate_per_min, emission_notifier_queue
             )
-        finally:
-            # Python convention: indicate that the previously enqueued task has been processed
-            emission_notifier_queue.task_done()
-
-
-async def emit(
-    mixnet: Mixnet,
-    redundancy: int,  # b in the spec
-    real_packet_queue: PacketQueue,
-    redundant_real_packet_queue: PacketQueue,
-    outbound_socket: PacketQueue,
-):
-    if not redundant_real_packet_queue.empty():
-        addr, packet = redundant_real_packet_queue.get_nowait()
-        await outbound_socket.put((addr, packet))
-        return
-
-    if not real_packet_queue.empty():
-        addr, packet = real_packet_queue.get_nowait()
-        # Schedule redundant real packets
-        for _ in range(redundancy - 1):
-            redundant_real_packet_queue.put_nowait((addr, packet))
-            await outbound_socket.put((addr, packet))
-
-    packet, route = PacketBuilder.drop_cover(b"drop cover", mixnet).next()
-    await outbound_socket.put((route[0].addr, packet))
-
-
-async def emission_notifier(emission_rate_per_min: int, queue: asyncio.Queue):
-    while True:
-        await asyncio.sleep(poisson_interval_sec(emission_rate_per_min))
-        queue.put_nowait(None)
+        )
+
+        while True:
+            # Wait until the next emission time
+            _ = await emission_notifier_queue.get()
+            try:
+                await self.__emit(self.__config.redundancy, redundant_real_packet_queue)
+            finally:
+                # Python convention: indicate that the previously enqueued task has been processed
+                emission_notifier_queue.task_done()
+
+    async def __emit(
+        self,
+        redundancy: int,  # b in the spec
+        redundant_real_packet_queue: PacketQueue,
+    ):
+        if not redundant_real_packet_queue.empty():
+            addr, packet = redundant_real_packet_queue.get_nowait()
+            await self.__outbound_socket.put((addr, packet))
+            return
+
+        if not self.__real_packet_queue.empty():
+            addr, packet = self.__real_packet_queue.get_nowait()
+            # Schedule redundant real packets
+            for _ in range(redundancy - 1):
+                redundant_real_packet_queue.put_nowait((addr, packet))
+                await self.__outbound_socket.put((addr, packet))
+
+        packet, route = PacketBuilder.drop_cover(
+            b"drop cover", self.__config.topology
+        ).next()
```

Comment:
I think the packet builder is a bit complicated to follow. Why in every instance we set the `self.iter` but is a Class attribute? Then it is wrapped in an iterator that just consumes a single packet/route?

Reply:
> Then it is wrapped in an iterator that just consumes a single packet/route?

For drop cover, yes, this class (iterator) returns only a single packet. For real traffic, the number of packets depends on the length of message provided.

> Why in every instance we set the self.iter but is a Class attribute? 

I thought it would be more readable to make PacketBuilder as an interator, but maybe it's not. I can probably inherit iterator, but I guess it would be much simpler to make it just return a list. Let me try and check which would be simpler.

Reply:
I just refactored it simpler: e0d9df1e3039fa8532fc2975720b739ac676cd71, instead of using iterator.